src/strategies/flex_window.py

```diff
@@ -84,170 +86,321 @@ def step(self, event_list=[]):
+                     if b.parent is not None]
+        cur_window = gc.window
+        sim_batteries = deepcopy(batteries)
+
+        # charge/discharge batteries
+        min_power = - gc.max_power
+        max_power = gc.max_power - gc.get_current_load()
+
+        window_timesteps = [
+            item for item in timesteps if item["window"] is cur_window
+        ]
+        new_timesteps = []
+        for i, row in enumerate(window_timesteps):
+            if window_timesteps[i]["timestep_idx"] != i:
+                break
+            new_timesteps.append(row)
+        old_soc = [b.soc for b in sim_batteries]
+
+        while max_power - min_power > self.EPS:
+            total_power = (min_power + max_power) / 2
+            # reset soc
+            for i, b in enumerate(sim_batteries):
+                b.soc = old_soc[i]
+
+            # calculate needed power to load battery
+            for ts_info in new_timesteps:
+                for b in sim_batteries:
+                    if cur_window:
+                        if b.soc > 1 - self.EPS:
+                            # adready charged
                             break
-                        if ts_info["window"] == charged_in_window:
-                            p = util.clamp_power(power, sim_vehicle, cs)
-                            avg_power = sim_vehicle.battery.load(self.interval, p)["avg_power"]
-                        elif not charged_in_window and ts_info["window"]:
-                            # charging windows not sufficient, charge max during window
-                            p = util.clamp_power(ts_info["power"], sim_vehicle, cs)
-                            avg_power = sim_vehicle.battery.load(self.interval, p)["avg_power"]
-
-                        power_vec[ts_idx] = avg_power
-                        safe = sim_vehicle.get_delta_soc() <= 0
-                        if safe:
-                            power_vec[ts_idx+1:] = [0]*(len(timesteps) - ts_idx - 1)
+                    else:
+                        if b.soc < 0 + self.EPS:
+                            # already discharged
                             break
 
-                    if safe:
-                        max_power = power
-                    else:
-                        min_power = power
+                    total_power = (
+                        0
+                        if total_power < b.min_charging_power
+                        else total_power
+                    )
+                    if total_power > 0:
+                        if cur_window:
+                            b.load(
+                                self.interval, (total_power / len(sim_batteries))
+                            )["avg_power"]
+                        else:
+                            b.unload(
+                                self.interval, (total_power / len(sim_batteries))
+                            )["avg_power"]
+            if cur_window:
+                safe = all(
```

Comment:
safe is probably the wrong word here, I'd recommend at_limit or something similar. Batteries are not critical for safety

Reply:
But in this case it really is about the charges vehicles "safe" to go, right? I thought you just meant that for batteries and v2g it is not critical, but for the charging functions it is